server/src/main/java/io/druid/client/indexing/IndexingServiceClient.java

```diff
@@ -76,25 +86,100 @@ public void mergeSegments(List<DataSegment> segments)
       }
     }
 
-    runQuery(new ClientAppendQuery(dataSource, segments));
+    postTask(new ClientAppendTask(dataSource, segments));
+  }
+
+  public String hadoopMergeSegments(
+      String dataSource,
+      List<Interval> intervalsToReindex,
+      AggregatorFactory[] aggregators,
+      QueryGranularity queryGranularity,
+      List<String> dimensions,
+      Map<String, Object> tuningConfig,
+      List<String> hadoopCoordinates
+  )
+  {
+    final InputStream queryResponse = postTask(
+        new ClientHadoopIndexTask(
+            String.format(
+                "%s_%s_%s",
+                DruidCoordinatorHadoopSegmentMerger.HADOOP_REINDEX_TASK_ID_PREFIX, dataSource, new DateTime()
+            ),
+            dataSource,
+            intervalsToReindex,
+            aggregators,
+            dimensions,
+            queryGranularity,
+            tuningConfig,
+            hadoopCoordinates,
+            jsonMapper
+        )
+    );
+
+    try {
+      return jsonMapper.<Map<String, String>>readValue(
+          queryResponse, new TypeReference<Map<String, String>>()
+          {
+          }
+      ).get("task");
+    }
+    catch (IOException e) {
+      log.warn(
+          "Error pasring HadoopReindexTask ID for dataSource [%s] and Interval [%s]",
+          dataSource,
+          intervalsToReindex
+      );
+      return null;
+    }
   }
 
   public void killSegments(String dataSource, Interval interval)
   {
-    runQuery(new ClientKillQuery(dataSource, interval));
+    postTask(new ClientKillTask(dataSource, interval));
   }
 
   public void upgradeSegment(DataSegment dataSegment)
   {
-    runQuery(new ClientConversionQuery(dataSegment));
+    postTask(new ClientConversionTask(dataSegment));
   }
 
   public void upgradeSegments(String dataSource, Interval interval)
   {
-    runQuery(new ClientConversionQuery(dataSource, interval));
+    postTask(new ClientConversionTask(dataSource, interval));
+  }
+
+  public List<Map<String, Object>> getIncompleteTasks()
+  {
+    final InputStream queryResponse = runGetQuery(INCOMPLETE_TASKS);
+    try {
+      return jsonMapper.readValue(
+          queryResponse, new TypeReference<List<Map<String, Object>>>()
+          {
+          }
+      );
+    }
+    catch (IOException e) {
+      log.warn("Error parsing response for incompleteTasks query from Overlord [%s]", baseUrl());
+      return null;
+    }
+  }
+
+  private InputStream runGetQuery(String path)
+  {
+    try {
+      return client.go(
+          new Request(
+              HttpMethod.GET,
+              new URL(String.format("%s/%s", baseUrl(), path))
+          ), RESPONSE_HANDLER
+      ).get();
```

Comment:
Isn't the default read timeout mechanism embedded in the http client sufficient for this case?